apps/openssl.c

```diff
@@ -216,6 +216,13 @@ static void setup_trace(const char *str)
 {
     char *val;
 
+    /*
+     * We add this handler as early as possible to ensure it's executed
+     * as late as possible, i.e. after the TRACE code has done its cleanup
+     * (which happens last in OPENSSL_cleanup).
+     */
+    atexit(cleanup_trace);
```

Comment:
Err. OPENSSL_cleanup is itself called by an atexit handler. AFAIK there is no guarantee about the ordering of different atexit handlers. So this might work on one platform but fail on another.

Reply:
The standard doesn't mention ordering.  It also specifies a minimum of 32 functions can be registered -- perhaps libcrypto could contain a stack of termination functions where the ordering can be controlled and the limit avoided?  It would also play nicer with the 32 lower limit.

Reply:
We already have such a stack. You create handlers via OPENSSL_atexit() - and then OPENSSL_cleanup() is responsible for running them. At the moment though it runs all the OpenSSL atexit handlers first, and then does other cleanup afterwards - and there is no possibility of ordering. Probably we ought to change it so that *all* cleanup is done via the OpenSSL atexit handlers and we add an ordering mechanism.

Reply:
This is what I meant, I should have been clearer :(

Reply:
https://en.cppreference.com/w/c/program/atexit

There is mention of ordering there (basically LIFO), and it refers to the C standards.
That's the way I have grown to know atexit()

Reply:
LGTM. But I‘d rather let Matt approve, because he expressed the concern.

Reply:
I checked C90, C99 and C11. I can't see any reference to the ordering in any of those. It appears to be a POSIX thing.

Reply:
The text specifying that is in the `exit` description.  Quoting from the C89 draft I currently have access to (`<=========` mine):

```
4.10.4.3 The exit function

Synopsis

         #include <stdlib.h>
         void exit(int status);

Description

   The exit function causes normal program termination to occur.  If
more than one call to the exit function is executed by a program, the
behavior is undefined.

   First, all functions registered by the atexit function are called,
in the reverse order of their registration./115/                    <=========

   Next, all open output streams are flushed, all open streams are
closed, and all files created by the tmpfile function are removed.

   Finally, control is returned to the host environment.  If the value
of status is zero or EXIT_SUCCESS , an implementation-defined form of
the status successful termination is returned.  If the value of status
is EXIT_FAILURE , an implementation-defined form of the status
unsuccessful termination is returned.  Otherwise the status returned
is implementation-defined.

Returns

   The exit function cannot return to its caller.  
```

I find similar descriptions in "7.22.4.4  The exit function" in http://open-std.org/JTC1/SC22/WG14/www/docs/n1570.pdf and in "7.20.4.3  The exit function" in http://open-std.org/JTC1/SC22/WG14/www/docs/n1256.pdf

If you have better standard papers, I would very much like to access them.

Reply:
Hmmmm.....you would think they would say that in the definition of atexit()! Strictly speaking the way its worded, it implies that ordering is only valid if "exit()" is explicitly called. So what happens if "main" just returns? But maybe I'm splitting hairs...so ok!

Reply:
The standards say that if main just returns it is equivalent to calling exit.

Reply:
(The intent is that if the program aborts, the handlers won't be called)

Reply:
What a maze of words and paper we abide by.